File: codebase/src/cpp/hla13/test/ddm/RequestUpdateWithRegionTest.cpp
/*
 *   Copyright 2008 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL)
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
#include "RequestUpdateWithRegionTest.h"

// Register test suite with the global repository
CPPUNIT_TEST_SUITE_REGISTRATION( RequestUpdateWithRegionTest );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RequestUpdateWithRegionTest, "RequestUpdateWithRegionTest" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RequestUpdateWithRegionTest, "requestUpdateWithRegion" );
CPPUNIT_TEST_SUITE_NAMED_REGISTRATION( RequestUpdateWithRegionTest, "ddm" );

/////////////////////////////////////////////////////////////////////////////////////////////
////////////////////////////////// Constructors/Destructors /////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////
RequestUpdateWithRegionTest::RequestUpdateWithRegionTest()
{
	this->defaultFederate = new Test13Federate( "defaultFederate" );
	this->provokerFederate = new Test13Federate( "provokerFederate" );
}

RequestUpdateWithRegionTest::~RequestUpdateWithRegionTest()
{
	delete this->provokerFederate;
	delete this->defaultFederate;
}

/////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////// Test Setup and Helper Methods ///////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////

void RequestUpdateWithRegionTest::setUp()
{
	this->defaultFederate->quickCreate();
	this->defaultFederate->quickJoin();
	this->provokerFederate->quickJoin();

	// get the handle information
	this->aHandle  = defaultFederate->quickOCHandle( "ObjectRoot.A" );
	this->aaHandle = defaultFederate->quickACHandle( "ObjectRoot.A", "aa" );
	this->abHandle = defaultFederate->quickACHandle( "ObjectRoot.A", "ab" );
	this->acHandle = defaultFederate->quickACHandle( "ObjectRoot.A", "ac" );
	this->bHandle  = defaultFederate->quickOCHandle( "ObjectRoot.A.B" );
	
	// publish and subscribe
	this->defaultFederate->quickPublish( "ObjectRoot.A", 2, "aa", "ab" );
	this->defaultFederate->quickPublish( "ObjectRoot.A.B", 3, "aa", "ab", "ac" );
	
	// create the regions
	this->senderRegion = defaultFederate->quickCreateTestRegion( 100, 200 );
	this->updateRegion = provokerFederate->quickCreateTestRegion( 150, 250 );
	this->updateRegionOOB = provokerFederate->quickCreateTestRegion( 1000, 2000 );
	
	// register some objects to work with
	this->firstObject = defaultFederate->quickRegisterWithRegion( "ObjectRoot.A", senderRegion,
	                                                              2, "aa", "ab" );
	this->secondObject = defaultFederate->quickRegisterWithRegion( "ObjectRoot.A.B", senderRegion,
	                                                              3, "aa", "ab", "ac" );
	
	// create the attribute handle set that is used for convenience
	this->ahs = defaultFederate->populatedAHS( 3, aaHandle, abHandle, acHandle );
}

void RequestUpdateWithRegionTest::tearDown()
{
	delete this->senderRegion;
	delete this->updateRegion;
	delete this->updateRegionOOB;
	delete this->ahs;
	
	this->provokerFederate->quickResign();
	this->defaultFederate->quickResign();
	this->defaultFederate->quickDestroy();
}

////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////// Test Methods ///////////////////////////////////////
////////////////////////////////////////////////////////////////////////////////////////////
// void requestClassAttributeValueUpdateWithRegion( RTI::ObjectClassHandle theClass,
//                                                  const RTI::AttributeHandleSet &theAttributes,
//                                                  const RTI::Region &theRegion )
//     throw( RTI::ObjectClassNotDefined,
//            RTI::AttributeNotDefined,
//            RTI::RegionNotKnown,
//            RTI::FederateNotExecutionMember,
//            RTI::ConcurrentAccessAttempted,
//            RTI::SaveInProgress,
//            RTI::RestoreInProgress,
//            RTI::RTIinternalError )

/////////////////////////////////////////////////
// TEST: (valid) testRequestUpdateWithRegion() //
/////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegion()
{
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( aHandle,
		                                                                      *ahs,
		                                                                      *updateRegion );
	}
	catch( RTI::Exception &e )
	{
		failTest( "Unexpected exception while requesting class update with ddm: %s", e._reason );
	}

	// make sure the callbacks get through to the federate that owns the attributes
	set<RTI::AttributeHandle>* firstAttributes =
		defaultFederate->fedamb->waitForProvideRequest( firstObject );
	CPPUNIT_ASSERT( 2 == firstAttributes->size() );
	CPPUNIT_ASSERT( firstAttributes->find(aaHandle) != firstAttributes->end() );
	CPPUNIT_ASSERT( firstAttributes->find(abHandle) != firstAttributes->end() );
	CPPUNIT_ASSERT( firstAttributes->find(acHandle) == firstAttributes->end() );
	delete firstAttributes;

	set<RTI::AttributeHandle>* secondAttributes =
		defaultFederate->fedamb->waitForProvideRequest( secondObject );
	CPPUNIT_ASSERT( secondAttributes->find(aaHandle) != firstAttributes->end() );
	CPPUNIT_ASSERT( secondAttributes->find(abHandle) != firstAttributes->end() );
	CPPUNIT_ASSERT( secondAttributes->find(acHandle) != firstAttributes->end() );
	delete secondAttributes;
}

//////////////////////////////////////////////////////////////////////////
// TEST: (valid) testRequestUpdateWithRegionUsingNonOverlappingRegion() //
//////////////////////////////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegionUsingNonOverlappingRegion()
{
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( aHandle,
		                                                                      *ahs,
		                                                                      *updateRegionOOB );
	}
	catch( RTI::Exception &e )
	{
		failTest( "Unexpected exception while requesting class update with ddm: %s", e._reason );
	}

	// make sure the callbacks DON'T get through to the federate (as the regions don't overlap)
	defaultFederate->fedamb->waitForProvideRequestTimeout( firstObject );
	defaultFederate->fedamb->waitForProvideRequestTimeout( secondObject );
}

////////////////////////////////////////////////////////////////
// TEST: testRequestUpdateWithRegionUsingInvalidObjectClass() //
////////////////////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegionUsingInvalidObjectClass()
{
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( 1000000,
		                                                                      *ahs,
		                                                                      *updateRegion );
		failTestMissingException( "ObjectClassNotDefined",
		                          "requesting update with invalid object class" );
	}
	catch( RTI::ObjectClassNotDefined &ocnd )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "ObjectClassNotDefined", e,
		                        "requesting update with invalid object class" );
	}
}

///////////////////////////////////////////////////////////////////
// TEST: testRequestUpdateWithRegionUsingInvalidAttributeClass() //
///////////////////////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegionUsingInvalidAttributeClass()
{
	RTI::AttributeHandleSet *badset = defaultFederate->populatedAHS( 100, aaHandle, abHandle );
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( aHandle,
		                                                                      *badset,
		                                                                      *updateRegion );
		delete badset;
		failTestMissingException( "AttributeNotDefined",
		                          "requesting update with invalid attribute handle" );
	}
	catch( RTI::AttributeNotDefined &attnd )
	{
		// success!
		delete badset;
	}
	catch( RTI::Exception &e )
	{
		delete badset;
		failTestWrongException( "AttributeNotDefined", e,
		                        "requesting update with invalid attribute handle" );
	}
}

///////////////////////////////////////////////////////////
// TEST: testRequestUpdateWithRegionUsingUnknownRegion() //
///////////////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegionUsingUnknownRegion()
{
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( aHandle,
		                                                                      *ahs,
		                                                                      *senderRegion );
		failTestMissingException( "RegionNotKnown",
		                          "requesting update with region created in other federate" );
	}
	catch( RTI::RegionNotKnown &rnk )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "RegionNotKnown", e, 
		                        "requesting update with region created in other federate" );
	}
}

//////////////////////////////////////////////////////
// TEST: testRequestUpdateWithRegionWhenNotJoined() //
//////////////////////////////////////////////////////
void RequestUpdateWithRegionTest::testRequestUpdateWithRegionWhenNotJoined()
{
	provokerFederate->quickResign();
	
	try
	{
		provokerFederate->rtiamb->requestClassAttributeValueUpdateWithRegion( aHandle,
		                                                                      *ahs,
		                                                                      *updateRegion );
		failTestMissingException( "FederateNotExecutionMember",
		                          "requesting update with region when not joined" );
	}
	catch( RTI::FederateNotExecutionMember &fnem )
	{
		// success!
	}
	catch( RTI::Exception &e )
	{
		failTestWrongException( "FederateNotExecutionMember", e, 
		                        "requesting update with region when not joined" );
	}
}

File: codebase/src/java/test/hlaunit/hla13/ddm/RequestUpdateWithRegionTest.java
/*
 *   Copyright 2007 The Portico Project
 *
 *   This file is part of portico.
 *
 *   portico is free software; you can redistribute it and/or modify
 *   it under the terms of the Common Developer and Distribution License (CDDL) 
 *   as published by Sun Microsystems. For more information see the LICENSE file.
 *   
 *   Use of this software is strictly AT YOUR OWN RISK!!!
 *   If something bad happens you do not have permission to come crying to me.
 *   (that goes for your lawyer as well)
 *
 */
package hlaunit.hla13.ddm;

import java.util.Set;

import hla.rti.AttributeHandleSet;
import hla.rti.AttributeNotDefined;
import hla.rti.FederateNotExecutionMember;
import hla.rti.ObjectClassNotDefined;
import hla.rti.RTIinternalError;
import hla.rti.Region;
import hla.rti.RegionNotKnown;
import hla.rti.RestoreInProgress;
import hla.rti.SaveInProgress;
import hlaunit.hla13.common.Abstract13Test;
import hlaunit.hla13.common.Test13Federate;

import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

@Test(singleThreaded=true, groups={"RequestUpdateWithRegionTest", "requestUpdate", "ddm"})
public class RequestUpdateWithRegionTest extends Abstract13Test
{
	//----------------------------------------------------------
	//                    STATIC VARIABLES
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                   INSTANCE VARIABLES
	//----------------------------------------------------------
	private Test13Federate provokerFederate;
	private int aHandle;
	private int bHandle;
	private int aaHandle;
	private int abHandle;
	private int acHandle;
	private AttributeHandleSet set;
	private int firstObject;
	private int secondObject;
	private Region senderRegion;
	private Region updateRegion;
	private Region updateRegionOOB;

	//----------------------------------------------------------
	//                      CONSTRUCTORS
	//----------------------------------------------------------

	//----------------------------------------------------------
	//                    INSTANCE METHODS
	//----------------------------------------------------------
	@BeforeClass(alwaysRun=true)
	public void beforeClass()
	{
		super.beforeClass();
		this.provokerFederate = new Test13Federate( "provokerFederate", this );
	}
	
	@BeforeMethod(alwaysRun=true)
	public void beforeMethod() throws Exception
	{
		defaultFederate.quickCreate();
		defaultFederate.quickJoin();
		provokerFederate.quickJoin();
		
		// cache the fom handles
		aHandle  = defaultFederate.quickOCHandle( "ObjectRoot.A" );
		aaHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "aa" );
		abHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "ab" );
		acHandle = defaultFederate.quickACHandle( "ObjectRoot.A", "ac" );
		bHandle  = defaultFederate.quickOCHandle( "ObjectRoot.A.B" );

		// register regions and do publish and subscribe
		defaultFederate.quickPublish( aHandle, aaHandle, abHandle, acHandle );
		defaultFederate.quickPublish( bHandle, aaHandle, abHandle, acHandle );
		senderRegion = defaultFederate.quickCreateTestRegion( 100, 200 );
		updateRegion = provokerFederate.quickCreateTestRegion( 150, 250 );
		updateRegionOOB = provokerFederate.quickCreateTestRegion( 1000, 2000 );
		
		// register an instance and discover it in the other federate //
		
		firstObject = defaultFederate.quickRegisterWithRegion( aHandle,
		                                                       senderRegion,
		                                                       aaHandle,
		                                                       abHandle );
		secondObject = defaultFederate.quickRegisterWithRegion( bHandle,
		                                                        senderRegion,
		                                                        aaHandle,
		                                                        abHandle,
		                                                        acHandle );

		// set up the default handle used during udpate
		set = provokerFederate.createAHS( aaHandle, abHandle, acHandle );
	}
	
	@AfterMethod(alwaysRun=true)
	public void afterMethod()
	{
		provokerFederate.quickResign();
		defaultFederate.quickResign();
		defaultFederate.quickDestroy();
	}
	
	@AfterClass(alwaysRun=true)
	public void afterClass()
	{
		super.afterClass();
	}

	//////////////////////////////////////////////////////////////////////////////////////////
	///////////////////////////////////// Helper Methods /////////////////////////////////////
	//////////////////////////////////////////////////////////////////////////////////////////

	//////////////////////////////////////////////////////////////////////////////////////////
	////////////////////////////// Request Class Update Methods //////////////////////////////
	//////////////////////////////////////////////////////////////////////////////////////////
	// public void requestClassAttributeValueUpdateWithRegion( int theClass,
	//                                                         AttributeHandleSet theAttributes,
	//                                                         Region theRegion )
	//        throws ObjectClassNotDefined,
	//               AttributeNotDefined,
	//               RegionNotKnown,
	//               FederateNotExecutionMember,
	//               SaveInProgress,
	//               RestoreInProgress,
	//               RTIinternalError,
	//               ConcurrentAccessAttempted;

	//////////////////////////////////////////////////////
	// TEST: (valid) testRequestClassUpdateWithRegion() //
	//////////////////////////////////////////////////////
	/**
	 * Two objects have been registered, one of type ObjectRoot.A and the other of type
	 * ObjectRoot.B. The provoker issues the request passing the class handle of ObjectRoot.B and
	 * all the handles for attributes from ObjectRoot.A.
	 * <p/>
	 * The defaultFederate should get a provide callback for the first object with only the first
	 * two attributes (as the third isn't associated with a region).
	 * <p/>
	 * The defaultFederate should get a provide callback for the second object with all three
	 * attributes.
	 */
	@Test
	public void testRequestClassUpdateWithRegion()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    updateRegion );
		}
		catch( Exception e )
		{
			unexpectedException( "Requesting class attribute update with region", e );
		}
		
		///////////////////////////////////////////////////////////////
		// validate that the requests reach the registering federate //
		///////////////////////////////////////////////////////////////
		Set<Integer> attributes = defaultFederate.fedamb.waitForProvideRequest( firstObject );
		Assert.assertEquals( attributes.size(), 2, "Was expecting request to update 2 handles" );
		Assert.assertTrue( attributes.contains(aaHandle),
		                   "Update request set didn't have aaHandle" );		
		Assert.assertTrue( attributes.contains(abHandle),
		                   "Update request set didn't have aaHandle" );		
		Assert.assertFalse( attributes.contains(acHandle),
		                    "Update request had abHandle (wasn't meant to)" );

		attributes = defaultFederate.fedamb.waitForProvideRequest( secondObject );
		Assert.assertEquals( attributes.size(), 3, "Was expecting request to update 3 handles" );
		Assert.assertTrue( attributes.contains(aaHandle),
		                   "Update request set didn't have aaHandle" );		
		Assert.assertTrue( attributes.contains(abHandle),
		                   "Update request set didn't have aaHandle" );		
		Assert.assertTrue( attributes.contains(acHandle),
		                   "Update request set didn't have aaHandle" );		
	}

	///////////////////////////////////////////////////////////////////////////////
	// TEST: (valid) testRequestClassUpdateWithRegionUsingNonOverlappingRegion() //
	///////////////////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingNonOverlappingRegion()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    updateRegionOOB );
		}
		catch( Exception e )
		{
			unexpectedException( "Requesting class attribute update with region", e );
		}
		
		// ensure the request reaches the other side
		defaultFederate.fedamb.waitForProvideRequestTimeout( firstObject );
		defaultFederate.fedamb.waitForProvideRequestTimeout( secondObject );
	}

	/////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingUndefinedClass() //
	/////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingUndefinedClass()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( 1000000,
			                                                                    set,
			                                                                    updateRegion );
			expectedException( ObjectClassNotDefined.class );
		}
		catch( ObjectClassNotDefined ocnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotDefined.class );
		}
	}

	////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingNegativeClass() //
	////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingNegativeClass()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( -1,
			                                                                    set,
			                                                                    updateRegion );
			expectedException( ObjectClassNotDefined.class );
		}
		catch( ObjectClassNotDefined ocnd )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, ObjectClassNotDefined.class );
		}
	}

	/////////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingUndefinedAttribute() //
	/////////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingUndefinedAttribute()
	{
		AttributeHandleSet badSet = provokerFederate.createAHS( aaHandle, abHandle, 1000000 );
		
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    badSet,
			                                                                    updateRegion );
			expectedException( AttributeNotDefined.class );
		}
		catch( AttributeNotDefined and )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, AttributeNotDefined.class );
		}
	}

	////////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingNegativeAttribute() //
	////////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingNegativeAttribute()
	{
		AttributeHandleSet badSet = provokerFederate.createAHS( aaHandle, abHandle, -1 );
		
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    badSet,
			                                                                    updateRegion );
			expectedException( AttributeNotDefined.class );
		}
		catch( AttributeNotDefined and )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, AttributeNotDefined.class );
		}
	}

	//////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithReginUsingNullAttibuteList() //
	//////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingNullAttributeList()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    null,
			                                                                    updateRegion );
			expectedException( RTIinternalError.class );
		}
		catch( RTIinternalError rtie )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RTIinternalError.class );
		}
	}

	////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingUnknownRegion() //
	////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingUnknownRegion()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    senderRegion );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnk )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}

	/////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionUsingNullRegion() //
	/////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionUsingNullRegion()
	{
		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle, set, null );
			expectedException( RegionNotKnown.class );
		}
		catch( RegionNotKnown rnk )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RegionNotKnown.class );
		}
	}

	////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionWhileNotJoined() //
	////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionWhileNotJoined()
	{
		provokerFederate.quickResign();

		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    updateRegion );
			expectedException( FederateNotExecutionMember.class );
		}
		catch( FederateNotExecutionMember fnem )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, FederateNotExecutionMember.class );
		}
	}

	////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionWhenSaveInProgress() //
	////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionWhenSaveInProgress()
	{
		provokerFederate.quickSaveInProgress( "save" );

		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    updateRegion );
			expectedException( SaveInProgress.class );
		}
		catch( SaveInProgress sip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, SaveInProgress.class );
		}
	}

	///////////////////////////////////////////////////////////////////
	// TEST: testRequestClassUpdateWithRegionWhenRestoreInProgress() //
	///////////////////////////////////////////////////////////////////
	@Test
	public void testRequestClassUpdateWithRegionWhenRestoreInProgress()
	{
		provokerFederate.quickRestoreInProgress( "save" );

		try
		{
			provokerFederate.rtiamb.requestClassAttributeValueUpdateWithRegion( aHandle,
			                                                                    set,
			                                                                    updateRegion );
			expectedException( RestoreInProgress.class );
		}
		catch( RestoreInProgress rip )
		{
			// success!
		}
		catch( Exception e )
		{
			wrongException( e, RestoreInProgress.class );
		}
	}

	//----------------------------------------------------------
	//                     STATIC METHODS
	//----------------------------------------------------------
}
